vankhoury/assignment1-AnnabelleAB
# Feature: Formula Evaluation & User Login Enhancement

### Summary:

This PR introduces several new features and bug fixes to improve the user experience and code reliability. 

#### New Features:
1. **FormulaEvaluator.ts**: Implemented new functionality to pass all given test scenarios.
2. **User Login**: Added a login function requiring users to enter a name before using the app's functionalities.

#### Bug Fixes:
1. **Test Cases**: Corrected several test cases to expect the correct outcomes.
2. **AC Button**: Fixed a bug where the AC (All Clear) button was not functioning as expected.

### Details:

#### FormulaEvaluator.ts
- Updated the `evaluate` function to handle special cases like division by zero, invalid formula, etc.
- Added new error messages and conditions for missing parentheses and other invalid formulas.
  
#### User Login
- Implemented a login function to require users to input their name.
- Only after logging in, users can access the app functionalities.

#### Test Cases
- Modified test cases that were failing or producing incorrect results.
- All tests now pass with expected outcomes.

#### AC Button
- Fixed the bug causing the AC button to malfunction.
- Now it correctly clears the current input and resets the state.

### How to test:
1. Run `npm test` to execute all the test cases.
2. Use the login functionality to see if it's requiring a username as expected.
3. Manually test the `AC` button to ensure that it's functioning correctly.

### Impact:
- User login provides a personalized user experience.
- Enhanced formula evaluation capabilities.
- Overall stability and reliability of the application.
